### ISSUE TITLE
Understanding generalization

### ISSUE THREAD

**Issue by kevinrobinson** [fairlearn/fairlearn#460]
hello!  in building on https://github.com/fairlearn/fairlearn/issues/444, I was curious what would happen if I took the quickstart example, split the data, and looked at how well the constraints held up on the test set.

Here's what I see in the training dataset using `ExponentiatedGradient` to satisfy the `DemographicParity` constraint on a `DecisionTreeClassifier`.  This is with a 70/30 train/test split, and shows the differences in these metrics between the predictions of the original model, and the predictions of the mitigated model:

```
TRAINING
   accuracy: 0.980 to 0.931
  selection: 0.230 to 0.275
   for male: 0.289 to 0.282
 for female: 0.110 to 0.261

TEST
   accuracy: 0.813 to 0.759
  selection: 0.233 to 0.292
   for male: 0.288 to 0.319
 for female: 0.119 to 0.236
```

This is just a single run with one seed, but the difference in demographic parity difference changes for selection rate changes quite a lot - from `0.021` (28.2% vs. 26.1%) to `0.083` (31.9% vs. 23.6%).

Of course in the real world there's distribution drift and many other things to consider, but I'm trying to understand even if the training and test data were drawn from the same underlying distribution with no noise, is the mitigation approach just overfitting?  Code for the modeling is below.

Setup:
```
X, y_true = adult()
trial_one = attempt(X, y_true, seed=11)
trial_two = attempt(X, y_true, seed=22)
```

Splitting, mitigating, and predicting:
```
from fairlearn.metrics import group_summary
from sklearn.metrics import accuracy_score
from sklearn.tree import DecisionTreeClassifier
from fairlearn.reductions import ExponentiatedGradient, DemographicParity, ConditionalSelectionRate
from sklearn.model_selection import train_test_split

def attempt(X, y, seed=42):
  np.random.seed(seed)
  (X_train, X_test, y_train, y_test) = train_test_split(X, y, test_size=0.30, random_state=seed)
  A_train = sex_for(X_train)
  A_test = sex_for(X_test)
  del X_train['Sex']
  del X_test['Sex']

  # train
  print('training...')
  (classifier, mitigator) = go(X_train, y_train, A_train)
  classifier.fit(X_train, y_train)

  # predict
  print('predicting...')
  y_train_predictions = classifier.predict(X_train)
  y_train_mitigated = mitigator.predict(X_train)

  # predict with mitigated
  print('predicting mitigated...')
  y_test_predictions = classifier.predict(X_test)
  y_test_mitigated = mitigator.predict(X_test)
  return (
    (X_train, y_train, A_train, y_train_predictions, y_train_mitigated),
    (X_test, y_test, A_test, y_test_predictions, y_test_mitigated)
  )

def sex_for(X):
  return X['Sex'].apply(lambda sex: "female" if sex == 0 else "male")

def go(X, y, A):
  constraint = DemographicParity()
  classifier = DecisionTreeClassifier()
  mitigator = ExponentiatedGradient(classifier, constraint)
  mitigator.fit(X, y, sensitive_features=A)
  return (classifier, mitigator)
```

Printing results:
```
from fairlearn import metrics

def printout(trial):
  train, test = trial
  (X_train, y_train, A_train, y_train_predictions, y_train_mitigated) = train
  (X_test, y_test, A_test, y_test_predictions, y_test_mitigated) = test
  output("TRAINING", y_train, y_train_predictions, y_train_mitigated, A_train)
  print('')
  output("TEST", y_test, y_test_predictions, y_test_mitigated, A_test)

def output(title, truth, predictions, mitigated, sensitive):
  print(title)
  print('   accuracy: {} to {}'.format(
    rate(accuracy_score(truth, predictions)),
    rate(accuracy_score(truth, mitigated)))
  )

  print('  selection: {} to {}'.format(
    rate(metrics.selection_rate(truth, predictions)),
    rate(metrics.selection_rate(truth, mitigated)))
  )
  predictions_summary = group_summary(metrics.selection_rate, truth, predictions, sensitive_features=sensitive)
  mitigated_summary = group_summary(metrics.selection_rate, truth, mitigated, sensitive_features=sensitive)
  print('   for male: {} to {}'.format(
    rate(predictions_summary['by_group']['male']),
    rate(mitigated_summary['by_group']['male'])
  ))
  print(' for female: {} to {}'.format(
    rate(predictions_summary['by_group']['female']),
    rate(mitigated_summary['by_group']['female'])
  ))

def rate(f):
  return '{0:.3f}'.format(f)
```

Thanks!

**Comment by romanlutz** [comment]
Interesting! Couple comments below:

- `DemographicParity` compares each individual group's selection rate to the overall selection rate, whereas demographic parity difference looks at min vs. max. This isn't document well and actually came up a few times already. Perhaps this isn't intuitive @MiroDudik 

- Ran this myself and got a warning in trial one `The pivot operation produces a pivot value of: 1.6e-06, which is only slightly greater than the specified tolerance 1.0e-09. This may lead to issues regarding the numerical stability of the simplex method. Removing redundant constraints, changing the pivot strategy via Bland's rule or increasing the tolerance may help reduce the issue.`

- Ran this with a few additional trials and the numbers are always around the same as yours, so it's not just that you had bad luck. For completeness:

```
>>> printout(trial_one)
TRAINING
   accuracy: 0.980 to 0.931
  selection: 0.230 to 0.275
   for male: 0.289 to 0.282
 for female: 0.110 to 0.261

TEST
   accuracy: 0.813 to 0.758
  selection: 0.233 to 0.293
   for male: 0.288 to 0.321
 for female: 0.119 to 0.235

>>> printout(trial_two)
TRAINING
   accuracy: 0.980 to 0.931
  selection: 0.224 to 0.269
   for male: 0.283 to 0.275
 for female: 0.104 to 0.256

TEST
   accuracy: 0.813 to 0.768
  selection: 0.241 to 0.286
   for male: 0.300 to 0.313
 for female: 0.122 to 0.230

>>> printout(trial_three)
TRAINING
   accuracy: 0.980 to 0.928
  selection: 0.227 to 0.273
   for male: 0.286 to 0.277
 for female: 0.107 to 0.264

TEST
   accuracy: 0.818 to 0.768
  selection: 0.232 to 0.283
   for male: 0.284 to 0.308
 for female: 0.124 to 0.232

>>> printout(trial_four)
TRAINING
   accuracy: 0.979 to 0.929
  selection: 0.224 to 0.269
   for male: 0.283 to 0.274
 for female: 0.105 to 0.258

TEST
   accuracy: 0.813 to 0.772
  selection: 0.243 to 0.285
   for male: 0.299 to 0.313
 for female: 0.127 to 0.227
```

- You ran `predict` which is randomized, so to validate that it's not just bad luck that way I ran it with `_pmf_predict` instead. The metrics had to be adjusted a little bit for that, too, since you now get a probability instead of 0/1. Results below for seeds 0 to 4:

```
# seed 0
TRAINING
   accuracy: 0.980 to 0.928
  selection: 0.226 to 0.272
   for male: 0.287 to 0.277
 for female: 0.103 to 0.262

TEST
   accuracy: 0.819 to 0.768
  selection: 0.234 to 0.290
   for male: 0.287 to 0.318
 for female: 0.128 to 0.233

# seed 1
TRAINING
   accuracy: 0.979 to 0.928
  selection: 0.231 to 0.277
   for male: 0.291 to 0.282
 for female: 0.111 to 0.267

TEST
   accuracy: 0.824 to 0.776
  selection: 0.227 to 0.277
   for male: 0.285 to 0.299
 for female: 0.113 to 0.235

# seed 2
TRAINING
   accuracy: 0.980 to 0.930
  selection: 0.223 to 0.270
   for male: 0.282 to 0.275
 for female: 0.105 to 0.260

TEST
   accuracy: 0.809 to 0.760
  selection: 0.238 to 0.291
   for male: 0.297 to 0.318
 for female: 0.118 to 0.237

# seed 3
TRAINING
   accuracy: 0.980 to 0.930
  selection: 0.225 to 0.269
   for male: 0.284 to 0.274
 for female: 0.107 to 0.259

TEST
   accuracy: 0.821 to 0.776
  selection: 0.239 to 0.287
   for male: 0.296 to 0.315
 for female: 0.122 to 0.231

# seed 4
TRAINING
   accuracy: 0.981 to 0.931
  selection: 0.227 to 0.271
   for male: 0.284 to 0.276
 for female: 0.110 to 0.261

TEST
   accuracy: 0.809 to 0.759
  selection: 0.242 to 0.288
   for male: 0.299 to 0.318
 for female: 0.127 to 0.227
```

Note: I had to modify the used metrics. Instead of `accuracy_score` I used `1 - mean_absolute_error` and `selection_rate` had to be modified to work with the floating point numbers as well (in a hacky way for this example as `sum(np.squeeze(np.asarray(y_pred))) / len(y_pred)`.

The results seem to rather reinforce your point, @kevinrobinson , so I'll leave this as is and will let @MiroDudik take a look.

**Comment by kevinrobinson** [comment]
@romanlutz thanks!  re: DemographicParity, yeah I moved your comment over to https://github.com/fairlearn/fairlearn/issues/444#issuecomment-641349702.

re: the error message, yes I saw that as well on the first run (see https://github.com/scipy/scipy/issues/11940) but ignored it since it's from a third party library.  I've been asking all these questions about one single narrowly technical use case (ie, the quick start), and think that seems good to start.

@MiroDudik's discussion on exponentiated gradient in https://github.com/fairlearn/fairlearn/issues/439 is helpful for putting these together, but it also makes me wonder if this work is all too early, and if any questions about generalization are still open research questions at this point.

**Comment by romanlutz** [comment]
Have you tried using different estimators? Perhaps the Decision Tree is just overfitting and the mitigation technique has nothing to do with it.

**Comment by kevinrobinson** [comment]
@romanlutz I took the test case verbatim from the quick start, and assumed that the quickstart scenario was making reasonable design choices and illustrating a situation where using fairlearn made sense.  If not then that's probably worth addressing first.

**Comment by MiroDudik** [comment]
It's definitely overfitting--you see that even on the level of non-constrained predictor. When I added reporting of confidence intervals on test data I got this for  `trial_one`:
```
TRAINING
   accuracy: 0.980 to 0.931
  selection: 0.230 to 0.275
   for male: 0.289 to 0.282
 for female: 0.110 to 0.261

TEST
   accuracy: 0.813±0.004 [n = 9769] to 0.759±0.004 [n = 9769]
  selection: 0.233±0.004 [n = 9769] to 0.292±0.005 [n = 9769]
   for male: 0.288±0.006 [n = 6550] to 0.319±0.006 [n = 6550]
 for female: 0.119±0.006 [n = 3219] to 0.236±0.007 [n = 3219]
```
After I modified the learner to:
```
classifier = DecisionTreeClassifier(min_samples_leaf=10, max_depth=4)
```
I got the following results:
```
>>> printout(modified_trial_one)
TRAINING
   accuracy: 0.843 to 0.826
  selection: 0.163 to 0.158
   for male: 0.214 to 0.163
 for female: 0.061 to 0.148

TEST
   accuracy: 0.847±0.004 [n = 9769] to 0.828±0.004 [n = 9769]
  selection: 0.158±0.004 [n = 9769] to 0.155±0.004 [n = 9769]
   for male: 0.208±0.005 [n = 6550] to 0.159±0.005 [n = 6550]
 for female: 0.055±0.004 [n = 3219] to 0.148±0.006 [n = 3219]

>>> printout(modified_trial_two)
TRAINING
   accuracy: 0.845 to 0.824
  selection: 0.160 to 0.166
   for male: 0.210 to 0.171
 for female: 0.057 to 0.156

TEST
   accuracy: 0.843±0.004 [n = 9769] to 0.824±0.004 [n = 9769]
  selection: 0.166±0.004 [n = 9769] to 0.172±0.004 [n = 9769]
   for male: 0.216±0.005 [n = 6551] to 0.177±0.005 [n = 6551]
 for female: 0.065±0.004 [n = 3218] to 0.163±0.007 [n = 3218]
```
So, we should definitely modify our quick start example! Thanks for catching this.

Overall, we should spell out in the documentation that just like regular classifiers can overfit, so can the mitigated classifiers and that hyperparameters should be tuned to prevent overfitting. As a strategy, we should suggest to tune hyperparameters on the "regular" classifier and then apply the same hyperparameters when running `ExponentiatedGradient`.

**Comment by kevinrobinson** [comment]
This is awesomely helpful, thanks!  I'll leave aside improvements to the quickstart, and try to focus on the conceptual questions so I can understand enough to roll this into a user guide.

First, I get that if the mitigation algorithm is given an overfit classifier, its predictions will be overfit and them mitigation algorithm will violate the bounds of the constraints.  This fundamentally happens because of the difference between the training and test sets.  But this surprised me, I think because I've made inaccurate assumptions about the "constraints" and "guarantees" that are provided.  My understanding is that if the training and test set are not drawn from the same distribution, there are no theoretical bounds on any of `fairlearn`'s constraints.  Perhaps this is obvious to everyone else, but this seems like it has practical implications that would be worth teaching users about :)

Second, I was wondering what an experiment might look like to probe at this.  I'll share this as a thought experiment first.  Imagine:
1. we made binary classifier that was perfect 95% of the time and random 5% of the time
2. we had a training dataset and a test dataset drawn from the same distribution
3. we ran a `ExponentiatedGradient` for `DemographicParity` on the messed up binary classifier from step 1
4. we looked at the predictions of the mitigated classifier on the test dataset

Would we be able to characterize the bounds of how much the `DemographicParity` constraint is violated on the test dataset, across all trials?  In other words, even if a toy scenario where we can control the error in the classifier, and the distribution drift between training and test, could we characterize the error bounds of `ExponentiatedGradient`?  Perhaps it's better to just make this example, and happy to do that if it'd be a more helpful next step!  I also recognize that if I more deeply understood the implementation code I would be able to explain this better, so if you all think it's helpful for me to go off and understand some particular piece of that, that could be a helpful next step too. 👍

**Comment by romanlutz** [comment]
@kevinrobinson I just submitted a PR to update the quickstart to avoid overfitting as described by Miro. That doesn't solve everything, obviously.

The guarantees only apply to the training set. The constraints will not hold on the test set, so a small difference is always expected. Furthermore, this is a randomized predictor as explained above. For more on that I recommend taking a look at the API reference https://fairlearn.github.io/api_reference/fairlearn.reductions.html#fairlearn.reductions.ExponentiatedGradient.predict as well as the paper's section on randomization (where they introduce `Q`).

We absolutely need to address that in the documentation. As you noticed there are currently two efforts by Richard and me to write documentation around basics of reductions approaches, and we should build on those PRs with this kind of info.

Regarding your thought experiment: Are you assuming that the "random 5%" are randomly picked once and from there on out consistent? I'm pretty sure Exponentiated Gradient assumes that the underlying estimators are deterministic.

**Comment by romanlutz** [comment]
The root of this issue still isn't addressed despite the correction in the Quickstart, of course!

One way to address it would be to have a notebook that illustrates generalization (or lack thereof) by taking something like this example (below is just a suggestion of what could be in that, not strictly the agenda):

1. overfit an estimator and see how Fairlearn's mitigation will overfit
2. add regularization (or other means) to avoid overfitting, and observe that Fairlearn's mitigation won't overfit either
3. discuss when fairness constraints are expected to hold, i.e., on the training set and if the test set follows the same distribution we expect to be reasonably close as well. This will probably also require mentioning randomized classifiers since that can affect results to some extent.

If anyone wants to pick this up just add a message below!
